Improve performance of drop table in iceberg connector

krvikash opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description
While working on Iceberg small files benchmarking. I noticed that the drop table is very slow when the number of files is too large for a table. The main reason for slowness is `dropTableData`, which deletes the files concurrently when the `fileIO` is not an instance of `SupportsBulkOperations`. we can delete files in bulk  `fileIO` is an instance of `SupportsBulkOperations`. So I introduced `BulkOperationsFileIO` which implements `SupportsBulkOperations`.

I tested it on my local machine. A table with glue metastore containing data files approx 10K took ~26 minutes without this change whereas with this change it only took ~20 seconds.

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(X) Release notes are required, with the following suggested text:

```markdown
# Section
* Improve performance of drop table in iceberg connector. ({issue}`issuenumber`)
```

ebyhr commented on lib/trino-filesystem/src/main/java/io/trino/filesystem/fileio/BulkOperationsFileIO.java:
```diff
@@ -0,0 +1,45 @@
+
+import java.io.IOException;
+
+public class BulkOperationsFileIO
```
```suggestion
public class BulkOperationsFileIo
```

ebyhr commented on lib/trino-filesystem/src/main/java/io/trino/filesystem/fileio/BulkOperationsFileIO.java:
```diff
@@ -0,0 +1,45 @@
+            throws BulkDeletionFailureException
+    {
+        ImmutableList.Builder<String> pathList = ImmutableList.builder();
+        pathsToDelete.forEach(path -> pathList.add(path));
```
nit: We could method reference.

ebyhr commented on lib/trino-filesystem/src/main/java/io/trino/filesystem/fileio/BulkOperationsFileIO.java:
```diff
@@ -0,0 +1,45 @@
+            fileSystem.deleteFiles(pathList.build());
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
```
Throwing `RuntimeException` is different from the interface's expectation. It would be nice to throw `BulkDeletionFailureException` or leave a comment.

krvikash replied:
`BulkDeletionFailureException` requires the number of files that failed to delete. Currently `fileSystem.deleteFiles` does not give such information. I will prepare a PR for the same.

ebyhr commented on lib/trino-filesystem/src/main/java/io/trino/filesystem/fileio/BulkOperationsFileIO.java:
```diff
@@ -0,0 +1,45 @@
+        ImmutableList.Builder<String> pathList = ImmutableList.builder();
+        pathsToDelete.forEach(path -> pathList.add(path));
+        try {
+            fileSystem.deleteFiles(pathList.build());
```
Don't we need to split into some batch likes `VacuumProcedure` in Delta Lake connector?

ebyhr commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -355,8 +357,10 @@ public void dropTable(ConnectorSession session, SchemaTableName schemaTableName)
-        dropTableData(table.io(), table.operations().current());
-        deleteTableDirectory(fileSystemFactory.create(session), schemaTableName, table.location());
+        TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+        BulkOperationsFileIO bulkOperationsFileIO = new BulkOperationsFileIO(fileSystem);
```
Same for other catalogs. 
```suggestion
        BulkOperationsFileIo bulkOperationsFileIo = new BulkOperationsFileIO(fileSystem);
```

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -355,8 +357,10 @@ public void dropTable(ConnectorSession session, SchemaTableName schemaTableName)
         catch (AmazonServiceException e) {
             throw new TrinoException(HIVE_METASTORE_ERROR, e);
         }
-        dropTableData(table.io(), table.operations().current());
```
Would it be possible to make `table.io()` return an instance of `SupportsBulkOperations`
so that we don't need to remember to wrap it in every place where it matters?

findepi commented on lib/trino-filesystem/src/main/java/io/trino/filesystem/fileio/BulkOperationsFileIo.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class BulkOperationsFileIo
+        extends ForwardingFileIo
```
I would implement this directly in ForwardingFileIo, not as a subclass.

(that would probably also satisfy https://github.com/trinodb/trino/pull/15981/files#r1097092879, right?)

cc @electrum

findepi commented on lib/trino-filesystem/src/main/java/io/trino/filesystem/fileio/BulkOperationsFileIo.java:
```diff
@@ -0,0 +1,62 @@
+        }
+        catch (IOException e) {
+            // TODO find out the number of files that failed to delete and throw BulkDeletionFailureException
+            throw new RuntimeException(e);
```
maybe something like

```
  throw new UncheckedIOException("Failed to delete some or all of files: " + paths, e);
```

findepi commented on lib/trino-filesystem/src/main/java/io/trino/filesystem/fileio/BulkOperationsFileIo.java:
```diff
@@ -0,0 +1,62 @@
+            fileSystem.deleteFiles(filesToDelete);
+        }
+        catch (IOException e) {
+            // TODO find out the number of files that failed to delete and throw BulkDeletionFailureException
```
what is `org.apache.iceberg.io.BulkDeletionFailureException#numberFailedObjects` needed for?

if we don't need it, we can skip having a TODO

findepi replied:
> SupportsBulkOperations::deleteFiles throws BulkDeletionFailureException exception.

I know. 
BTW it can also throw any unckeched exception, obviously.

> numberFailedObjects which means the number of files that failed to delete.

I figured this from the name :)
How does Iceberg lib use that information?

homar commented on lib/trino-filesystem/src/main/java/io/trino/filesystem/fileio/BulkOperationsFileIo.java:
```diff
@@ -0,0 +1,62 @@
+    public void deleteFiles(Iterable<String> pathsToDelete)
+            throws BulkDeletionFailureException
+    {
+        List<String> filesToDelete = new ArrayList<>();
```
Please use `ImmutableList.Builder` instead of `ArrayList`.
Also I think you can replace this logic here with something like:

```
Iterable<List<String>> partitions = Iterables.partition(pathsToDelete, DELETE_BATCH_SIZE);
for (List<String> currentBatch : partitions) {
    deleteBatch(currentBatch);
}
```

findepi replied:
`ArrayList` better over `ImmutableList.Builder` because of re-allocations (https://github.com/trinodb/trino/pull/14434/files#diff-e1cb17efec6787989f9df9ee40c4f2809ff3fe946cd2ec721ff8932b131997b8R1183)

👍 for `Iterables.partition`

krvikash commented:
Addressed comments.

CI is failing due to a flaky test

> 2023-02-06T12:27:38.4496953Z tests               | 2023-02-06 18:12:38 INFO: FAILURE     /    io.trino.tests.product.kudu.TestKuduConnectoKerberosSmokeTest.kerberosAuthTicketExpiryTest (Groups: profile_specific_tests, kudu) took 3.3 seconds

https://github.com/trinodb/trino/issues/14441

electrum commented on lib/trino-filesystem/src/main/java/io/trino/filesystem/fileio/ForwardingFileIo.java:
```diff
@@ -61,4 +66,22 @@ public void deleteFile(String path)
+
+    @Override
+    public void deleteFiles(Iterable<String> pathsToDelete)
+            throws BulkDeletionFailureException
```
This isn’t thrown by the method. Should we be using this instead?

findepi replied:
we cannot throw BulkDeletionFailureException becauase we cannot fill `BulkDeletionFailureException#numberFailedObjects`

see https://github.com/trinodb/trino/pull/15981#discussion_r1097111381 for more

ebyhr commented on lib/trino-filesystem/src/main/java/io/trino/filesystem/fileio/ForwardingFileIo.java:
```diff
@@ -61,4 +66,22 @@ public void deleteFile(String path)
+            fileSystem.deleteFiles(filesToDelete);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException("Failed to delete some or all of files: " + filesToDelete, e);
```
I don't think including all file paths to the error message is a good idea. It may include 1,000 files in large deletes.

krvikash replied:
Hi  @ebyhr| @findepi | @electrum , I am proposing a PR https://github.com/trinodb/trino/pull/16006 to track the number of files that failed to delete. If https://github.com/trinodb/trino/pull/16006 gets approved then we can throw `BulkDeletionFailureException` over here.

findepi replied:
@ebyhr you convinced me, let's not make the message that long. OTOH, it could be problematic if we do not include any paths in the message. It can make identifying problems harder (eg which bucket did we try to access?).
I pushed a change to display only first 5 paths and then ellipsis `...`.
